### PR TITLE
Introduce Utilities::MemorySpace namespace and MemoryBlock class

### PR DESCRIPTION
**Pull request by Rombur**
The goal of this PR is to make it easier to write code that works both on the host and the device. For that reason the PR does the following:
 1. Add helper functions to allocate memory, free memory, and copy memory on the host and the device. I use tag dispatching to choose the correct implement, i.e, allocate memory uses `new` on the host but `cudaMalloc` on the device
 2. Add `for_each` function that performs a simple for loop on the host and launches a kernel on the device
 3. Right now the only data structure that works on the host and the device is `LA::distributed::Vector`. I have added a new class `MemoryBlock` that allocate a block of memory either on the host or the device. To access the underlying data of this class you need to use an `ArrayView`. I've done this because of limitations of lambda function with CUDA. In particular, you cannot use a private member data in the lambda. Since you have to use an `ArrayView` to access the data, you can safely have a private `MemoryBlock` and then, create an `ArrayView` just before calling the lambda. The other advantage is that the copy constructor of `MemoryBlock` is a regular copy constructor, if `MemoryBlock` could be used directly in a kernel, we would have to do a shallow copy and then keep track of the number of copies.

### CODE REVIEW

**Comment by Rombur** [comment]
Is there anything else blocking this PR?  Am I fine closing this if we decide that we are moving to Kokkos but in that case, I would like to have more input on things like design, scope of the work, etc.

File: include/deal.II/base/memory_space_utils.h [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+   */
+  namespace MemorySpace
+  {
+#ifdef DEAL_II_COMPILER_CUDA_AWARE
+    /**
+     * Allocate memory on the device.
+     */
+    template <typename Number>
+    Number *
+    allocate_data(const std::size_t size, const ::dealii::MemorySpace::CUDA &)
```
[the remark this answers is not in the document]
Comment (bangerth) [reply include/deal.II/base/memory_space_utils.h:42]: Yes, I'm just asking for it to be spelled out in the documentation :-)

**Comment by bangerth** [comment]
We already recognize Kokkos in cmake, though I have to admit that it's not clear to me what for. I see no reason not to already now *require* Kokkos for CUDA code. I wouldn't be opposed to requiring Kokkos in general if that is necessary to make everyone's life easier.

I'm not an expert in Kokkos, so I'm not sure what I can help with. But I'm happy to learn something about it if you'd like everyone interested in it to participate in a Zoom call. When you say "more input", what are you specifically looking for? I assume you're hoping for something you can also tell your management?

**Comment by masterleinad** [comment]
I left a comment in #12894 and I'm also happy with having a Zoom call for this.

**Comment by Rombur** [comment]
> We already recognize Kokkos in cmake, though I have to admit that it's not clear to me what for. 

We use Kokkos for ArborX, the catch is that it only works on the CPU. This avoids some of the ugliness of using Kokkos with nvcc.

>  When you say "more input", what are you specifically looking for? 

There are two things I am looking for:
 1. some technical feedback on how the new interface should look like but I would expect that only @masterleinad knows enough about Kokkos to help
 2. buy-in from the other developers. So far, people who didn't use CUDA could basically pretends that the CUDA code didn't exist. This will be harder to do with Kokkos. I don't want to get into a situation where some code needs to be changed/needs to use Kokkos because of CUDA but the PR is rejected because someone refuses to have Kokkos in that part of the code. We could keep Kokkos separated the way CUDA is right now, but to me that really decreases the interest of using Kokkos.

**Comment by bangerth** [comment]
About 2, let's talk about it tomorrow. I tend to think that we all have to learn something like Kokkos at some point, and this might be the point.

**Comment by drwells** [comment]
Since we now require Kokkos, do we still need this patch?

**Pull request closed without merging**